bindings/c/include/opendal.h

```diff
@@ -25,14 +25,229 @@
+ Note that if `error` is not null, the error consumes heap memory
+ remember to free it by using [`opendal_free_error`]
+ */
+typedef struct opendal_result_write {
+  bool success;
+  struct opendal_error *error;
+} opendal_result_write;
+
+/*
+ The [`Bytes`] type is a C-compatible substitute for [`Bytes`]
+ in Rust, it will not be deallocated automatically like what
+ has been done in Rust. Instead, you have to call [`opendal_free_bytes`]
+ to free the heap memory to avoid memory leak.
+ The field `data` should not be modified since it might causes
+ the reallocation of the Vector.
+ */
+typedef struct opendal_bytes {
+  const uint8_t *data;
+  uintptr_t len;
+} opendal_bytes;
+
+/*
+ The Rust-like Result type of opendal C binding, it contains
+ the data that the read operation returns and a pointer to an error type.
+ If the read operation failed, the `data` fields should be a nullptr.
+ Note that if `error` is not null, the error consumes heap memory
+ remember to free it by using [`opendal_free_error`]
+ */
+typedef struct opendal_result_read {
+  struct opendal_bytes *data;
+  struct opendal_error *error;
+} opendal_result_read;
+
 #ifdef __cplusplus
 extern "C" {
 #endif // __cplusplus
 
 /*
- Hello, OpenDAL!
+ Returns a result type [`opendal_result_op`], with operator_ptr. If the construction succeeds
+ the error is nullptr, otherwise it contains the error information.
+
+ # Safety
+
+ It is [safe] under two cases below
+ * The memory pointed to by `scheme` must contain a valid nul terminator at the end of
+   the string.
+ * The `scheme` points to NULL, this function simply returns you a null opendal_operator_ptr
+ */
+struct opendal_result_op opendal_new_operator(const char *scheme);
+
+/*
+ Write the data into the path blockingly by operator, returns whether the write succeeds
+ with error type
+
+ # Safety
+
+ It is [safe] under two cases below
+ * The memory pointed to by `path` must contain a valid nul terminator at the end of
+   the string.
+ * The `path` points to NULL, this function simply returns you false
+ */
+struct opendal_result_write opendal_operator_blocking_write(opendal_operator_ptr op_ptr,
+                                                            const char *path,
+                                                            struct opendal_bytes bytes);
+
+/*
+ Read the data out from path into a [`Bytes`] blockingly by operator, returns
+ a result with error type
+
+ # Safety
+
+ It is [safe] under two cases below
+ * The memory pointed to by `path` must contain a valid nul terminator at the end of
+   the string.
+ * The `path` points to NULL, this function simply returns you a nullptr
+ */
+struct opendal_result_read opendal_operator_blocking_read(opendal_operator_ptr op_ptr,
+                                                          const char *path);
+
+/*
+ Free the [`CError`] type's heap memory, it is safe to be called
+ on a nullptr
+ */
+void opendal_free_error(struct opendal_error *self);
```

Comment:
How about we stick with the naming convention of `opendal_<noun>_<verb>`? It seems more straightforward and easier to understand.

Reply:
Do you mean `opendal_error_free`? I am ok with it but I want to say that

`opendal_free_error` means **free** the **error**,
`opendal_operator_blocking_read` means use the **operator** to **read**

What's your opinion?

Reply:
> What's your opinion?

I still like `opendal_error_free`. But as discussed in https://github.com/apache/incubator-opendal/pull/1861#discussion_r1159983964, I prefer to remove `opendal_error` first.